File: packages/xyd-gql/src/opendocs.ts
import {
    GraphQLSchema,
    GraphQLArgument,
    GraphQLEnumType,
    GraphQLEnumValue,
    GraphQLField,
    GraphQLInputObjectType,
    GraphQLInterfaceType,
    GraphQLObjectType,
    GraphQLScalarType,
    GraphQLUnionType,
    ConstValueNode,
    StringValueNode,
} from "graphql";

import {GQLOperation, GQLSchemaToReferencesOptions, SortItem} from "./types";
import {GraphqlUniformReferenceType} from "./gql-core";
import {Context} from "./context";

export const OPEN_DOCS_SCHEMA_DIRECTIVE_NAME = "docs";
export const OPEN_DOCS_DIRECTIVE_NAME = "doc";

export function openDocsExtensionsToOptions(
    schema: GraphQLSchema,
) {
    const options: GQLSchemaToReferencesOptions = {}

    // Check for @docs directive in schema extensions
    for (const extension of schema.extensionASTNodes || []) {
        if (extension.kind === 'SchemaExtension') {
            for (const directive of extension.directives || []) {
                if (directive.name.value === OPEN_DOCS_SCHEMA_DIRECTIVE_NAME) {
                    for (const arg of directive.arguments || []) {
                        if (arg.name.value === 'flattenTypes' && arg.value.kind === 'BooleanValue') {
                            if (arg.value.value === true) {
                                options.flat = true
                            } else if (arg.value.value === false) {
                                options.flat = false
                            }
                        } else if (arg.name.value === 'sort' && arg.value.kind === 'ListValue') {
                            const sortItems: SortItem[] = [];
                            for (const item of arg.value.values) {
                                if (item.kind === 'ObjectValue') {
                                    const sortItem: SortItem = {};
                                    for (const field of item.fields) {
                                        if (field.name.value === 'node' && field.value.kind === 'StringValue') {
                                            sortItem.node = field.value.value;
                                        } else if (field.name.value === 'group' && field.value.kind === 'ListValue') {
                                            sortItem.group = field.value.values
                                                .filter((v): v is StringValueNode => v.kind === 'StringValue')
                                                .map(v => v.value);
                                        } else if (field.name.value === 'stack' && field.value.kind === 'IntValue') {
                                            sortItem.stack = parseInt(field.value.value);
                                        }
                                    }
                                    sortItems.push(sortItem);
                                }
                            }
                            if (!options.sort) {
                                options.sort = {};
                            }
                            options.sort.sort = sortItems;
                        } else if (arg.name.value === 'sortStack' && arg.value.kind === 'ListValue') {
                            const sortStacks: string[][] = [];
                            for (const item of arg.value.values) {
                                if (item.kind === 'ListValue') {
                                    const stack = item.values
                                        .filter((v): v is StringValueNode => v.kind === 'StringValue')
                                        .map(v => v.value);
                                    sortStacks.push(stack);
                                }
                            }
                            if (!options.sort) {
                                options.sort = {};
                            }
                            options.sort.sortStack = sortStacks;
                        } else if (arg.name.value === 'route' && arg.value.kind === 'StringValue') {
                            options.route = arg.value.value;
                        }
                    }
                }
            }
        }
    }

    return options;
}

type OpenDocsGQLNode =
    | GQLOperation
    | GraphQLScalarType
    | GraphQLObjectType
    | GraphQLField<any, any>
    | GraphQLArgument
    | GraphQLInterfaceType
    | GraphQLUnionType
    | GraphQLEnumType
    | GraphQLEnumValue
    | GraphQLInputObjectType

export function openDocsToGroup(
    ctx: Context | undefined,
    odGqlNode: OpenDocsGQLNode,
): string[] {
    let groups: string[] = []

    const metadata = (ctx?.schema as any).__metadata;

    if (metadata?.rootGroups) {
        groups = [...metadata.rootGroups];
    }

    let directiveGroups = false

    // Check schema metadata for field-specific groups (for operations)
    if (ctx?.schema && 'name' in odGqlNode) {
        const metadata = (ctx.schema as any).__metadata;
        if (metadata?.fields) {
            if ("_operationType" in odGqlNode) {
                let fieldKey = ""

                switch (odGqlNode._operationType) {
                    case "query": {
                        fieldKey = `Query.${odGqlNode.name}`;
                        break;
                    }
                    case "mutation": {
                        fieldKey = `Mutation.${odGqlNode.name}`;
                        break;
                    }
                    case "subscription": {
                        fieldKey = `Subscription.${odGqlNode.name}`;
                        break;
                    }
                }

                const fieldMetadata = metadata.fields.get(fieldKey);
                if (fieldMetadata?.groups) {
                    directiveGroups = true
                    groups.push(...fieldMetadata.groups);
                }
            }
        }
    }

    // If no groups from metadata, try getting groups from the node itself
    if (!directiveGroups && odGqlNode.astNode?.directives) {
        for (const directive of odGqlNode.astNode.directives) {
            switch (directive.name.value) {
                case OPEN_DOCS_DIRECTIVE_NAME: {
                    const groupArg = directive.arguments?.find((arg: {
                        name: { value: string }
                    }) => arg.name.value === 'group')
                    if (groupArg?.value.kind === 'ListValue') {
                        directiveGroups = true
                        groups.push(...groupArg.value.values
                            .filter((v: ConstValueNode): v is StringValueNode => v.kind === 'StringValue')
                            .map(v => v.value)
                        )
                    }
                    break
                }
            }
        }
    }

    // If still no groups, use default based on type
    if (!directiveGroups) {
        if (odGqlNode instanceof GraphQLObjectType) {
            groups.push("Objects")
        } else if (odGqlNode instanceof GraphQLInterfaceType) {
            groups.push("Interfaces")
        } else if (odGqlNode instanceof GraphQLUnionType) {
            groups.push("Unions")
        } else if (odGqlNode instanceof GraphQLEnumType) {
            groups.push("Enums")
        } else if (odGqlNode instanceof GraphQLInputObjectType) {
            groups.push("Inputs")
        } else if (odGqlNode instanceof GraphQLScalarType) {
            groups.push("Scalars")
        } else if (odGqlNode instanceof GQLOperation) {
            switch (odGqlNode._operationType) {
                case "query": {
                    groups.push("Queries")
                    break;
                }
                case "mutation": {
                    groups.push("Mutations")
                    break;
                }
                case "subscription": {
                    groups.push("Subscriptions")
                    break;
                }
            }
        }
    }

    return groups
}

export function openDocsCanonical(
    ctx: Context,
    gqlType: GraphqlUniformReferenceType,
) {
    let path = ""

    // Get parent path if this is a field
    if ('astNode' in gqlType && gqlType.astNode?.kind === 'FieldDefinition' && ctx?.schema) {
        // Check schema metadata for field-specific path
        const metadata = (ctx.schema as any).__metadata;
        if (metadata?.fields && 'name' in gqlType) {
            if ("_operationType" in gqlType) {
                let fieldKey = ""

                switch (gqlType._operationType) {
                    case "query": {
                        fieldKey = `Query.${gqlType.name}`;
                        break;
                    }
                    case "mutation": {
                        fieldKey = `Mutation.${gqlType.name}`;
                        break;
                    }
                    case "subscription": {
                        fieldKey = `Subscription.${gqlType.name}`;
                        break;
                    }
                }
                const fieldMetadata = metadata.fields.get(fieldKey);


                if (fieldMetadata?.path) {
                    path = fieldMetadata.path;
                }
            }
        }
    }

    // Extract path from @doc directive if present
    if (!path && gqlType.astNode?.directives) {
        for (const directive of gqlType.astNode.directives) {
            if (directive.name.value === OPEN_DOCS_DIRECTIVE_NAME) {
                const pathArg = directive.arguments?.find(arg => arg.name.value === 'path')
                if (pathArg?.value.kind === 'StringValue') {
                    path = pathArg.value.value
                }
            }
        }
    }

    return path
}

File: crates/xyd_gql/src/opendocs.rs
//! OpenDocs `@docs`/`@doc` directive semantics — port of `src/opendocs.ts`.

use crate::model::{
    directive_bool, directive_scope_values, directive_string, directive_string_list, FieldDef,
    Kind, SchemaModel, ScopeRef, TypeDef,
};
use crate::{Options, SortItem};
use async_graphql_value::ConstValue;

/// The node kinds `uniformify` distinguishes (port of the JS instanceof chain).
#[derive(Debug, Clone)]
pub enum Node<'a> {
    Type(&'a TypeDef),
    Operation { op: &'a str, field: &'a FieldDef },
    Field(&'a FieldDef),
}

/// `openDocsExtensionsToOptions`: read `extend schema @docs(...)`.
pub fn extensions_to_options(model: &SchemaModel) -> Options {
    let mut options = Options::default();
    for d in &model.schema_extension_directives {
        if d.name.node != "docs" {
            continue;
        }
        if let Some(flat) = directive_bool(d, "flattenTypes") {
            options.flat = Some(flat);
        }
        if let Some(v) = d.get_argument("sort") {
            if let ConstValue::List(items) = &v.node {
                let mut sort_items = Vec::new();
                for item in items {
                    let ConstValue::Object(fields) = item else {
                        continue;
                    };
                    let mut si = SortItem::default();
                    for (k, fv) in fields {
                        match (k.as_str(), fv) {
                            ("node", ConstValue::String(s)) => si.node = Some(s.clone()),
                            ("group", ConstValue::List(gs)) => {
                                si.group = Some(
                                    gs.iter()
                                        .filter_map(|g| match g {
                                            ConstValue::String(s) => Some(s.clone()),
                                            _ => None,
                                        })
                                        .collect(),
                                )
                            }
                            ("stack", ConstValue::Number(n)) => {
                                si.stack = n.as_i64().map(|i| i as usize)
                            }
                            _ => {}
                        }
                    }
                    sort_items.push(si);
                }
                options.sort.get_or_insert_with(Default::default).sort = Some(sort_items);
            }
        }
        if let Some(v) = d.get_argument("sortStack") {
            if let ConstValue::List(items) = &v.node {
                let mut stacks = Vec::new();
                for item in items {
                    if let ConstValue::List(vals) = item {
                        stacks.push(
                            vals.iter()
                                .filter_map(|g| match g {
                                    ConstValue::String(s) => Some(s.clone()),
                                    _ => None,
                                })
                                .collect(),
                        );
                    }
                }
                options.sort.get_or_insert_with(Default::default).sort_stack = Some(stacks);
            }
        }
        if let Some(route) = directive_string(d, "route") {
            options.route = Some(route);
        }
    }
    options
}

/// `openDocsToGroup` — groups for a node: rootGroups + (operation metadata ||
/// `@doc(group:)` on the node) || kind default.
pub fn to_group(model: &SchemaModel, node: &Node) -> Vec<String> {
    let mut groups: Vec<String> = model.root_groups.clone().unwrap_or_default();
    let mut directive_groups = false;

    // Operation field metadata (from docDirectiveChain).
    if let Node::Operation { op, field } = node {
        let key = format!("{}.{}", op_root_name(op), field.name);
        if let Some(md) = model.field_metadata.get(&key) {
            if let Some(g) = &md.groups {
                directive_groups = true;
                groups.extend(g.clone());
            }
        }
    }

    // `@doc(group:)` directly on the node.
    if !directive_groups {
        let directives = match node {
            Node::Type(t) => &t.directives,
            Node::Operation { field, .. } => &field.directives,
            Node::Field(f) => &f.directives,
        };
        for d in directives {
            if d.name.node == "doc" {
                if let Some(g) = directive_string_list(d, "group") {
                    directive_groups = true;
                    groups.extend(g);
                }
            }
        }
    }

    // Kind defaults.
    if !directive_groups {
        match node {
            Node::Type(t) => match t.kind {
                Kind::Object => groups.push("Objects".into()),
                Kind::Interface => groups.push("Interfaces".into()),
                Kind::Union => groups.push("Unions".into()),
                Kind::Enum => groups.push("Enums".into()),
                Kind::Input => groups.push("Inputs".into()),
                Kind::Scalar => groups.push("Scalars".into()),
            },
            Node::Operation { op, .. } => match *op {
                "query" => groups.push("Queries".into()),
                "mutation" => groups.push("Mutations".into()),
                "subscription" => groups.push("Subscriptions".into()),
                _ => {}
            },
            Node::Field(_) => {}
        }
    }

    groups
}

/// `openDocsCanonical` — metadata path (operations) or `@doc(path:)` on the node.
pub fn canonical(model: &SchemaModel, node: &Node) -> String {
    if let Node::Operation { op, field } = node {
        let key = format!("{}.{}", op_root_name(op), field.name);
        if let Some(md) = model.field_metadata.get(&key) {
            if let Some(p) = &md.path {
                return p.clone();
            }
        }
    }
    let directives = match node {
        Node::Type(t) => &t.directives,
        Node::Operation { field, .. } => &field.directives,
        Node::Field(f) => &f.directives,
    };
    for d in directives {
        if d.name.node == "doc" {
            if let Some(p) = directive_string(d, "path") {
                return p;
            }
        }
    }
    String::new()
}

/// `extractScopesFromDocDirective` — `@doc(scopes: [ENUM | "literal"])`;
/// enum refs resolve through `OpenDocsScope`'s `@scope(value:)`.
pub fn scopes(model: &SchemaModel, node: &Node) -> Vec<String> {
    let directives = match node {
        Node::Type(t) => &t.directives,
        Node::Operation { field, .. } => &field.directives,
        Node::Field(f) => &f.directives,
    };
    let mut out = Vec::new();
    for d in directives {
        if d.name.node != "doc" {
            continue;
        }
        for scope in directive_scope_values(d, "scopes") {
            match scope {
                ScopeRef::Literal(s) => out.push(s),
                ScopeRef::Enum(value_name) => {
                    // The current type if it's an enum, else OpenDocsScope.
                    let enum_type = match node {
                        Node::Type(t) if t.kind == Kind::Enum => Some(*t),
                        _ => model.get_type("OpenDocsScope"),
                    };
                    if let Some(et) = enum_type {
                        if let Some(ev) = et.values.iter().find(|v| v.name == value_name) {
                            for ed in &ev.directives {
                                if ed.name.node == "scope" {
                                    if let Some(v) = directive_string(ed, "value") {
                                        out.push(v);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    out
}

fn op_root_name(op: &str) -> &'static str {
    match op {
        "query" => "Query",
        "mutation" => "Mutation",
        "subscription" => "Subscription",
        _ => "Query",
    }
}

/// Internal opendocs types excluded from output (port of `isInternalOpenDocsType`).
pub fn is_internal_opendocs_type(name: &str) -> bool {
    matches!(
        name,
        "OpenDocsScope"
            | "OpenDocsSidebarItemType"
            | "OpenDocsPage"
            | "OpenDocsExampleInput"
            | "OpenDocsSortInput"
            | "OpenDocsSortStackInput"
    )
}

/// graphql-js built-in scalars + introspection (port of `isBuiltInType`).
pub fn is_built_in_type(name: &str) -> bool {
    matches!(name, "String" | "Int" | "Float" | "Boolean" | "ID") || name.starts_with("__")
}
